### PR TITLE
Feature/algolia places provider

### PR DESCRIPTION
**Pull request by sheub**
Hi @Fabex ,

The tests for the AlgoliaPlaces provider.
The tests have passed the Travis build. Still, I haven't run them with real API_KEY/APP_ID.

Also, I had to add a copy of the function `getParsedResponse` to AlgoliaPlaces.php otherwise Travis throw an error...
Sheub

### CODE REVIEW

File: src/Provider/AlgoliaPlaces/AlgoliaPlaces.php [hunk 1]
```diff
@@ -72,20 +75,34 @@ public function getName(): string
...
 
-        $jsonResponse = json_decode($this->getUrlContents(self::ENDPOINT_URL_SSL));
+        $request = $this->getRequest(self::ENDPOINT_URL_SSL);
+        $jsonResponse = json_decode($this->getParsedResponse($request), true);
+        // $jsonResponse = json_decode($this->getUrlContents(self::ENDPOINT_URL_SSL), true);
 
+        // error_log(json_encode($jsonResponse));
         if (is_null($jsonResponse)) {
             return new AddressCollection([]);
+        } elseif ($jsonResponse['degradedQuery']) {
```
Comment (Fabex) [inline src/Provider/AlgoliaPlaces/AlgoliaPlaces.php]: no need elseif and else. you have a return in your condition you can just do

```php
if() {return ...;}
if() {return ...;}
if() {return ...;}
return ...;
```

**Review by Fabex — CHANGES_REQUESTED** [review]

File: src/Provider/AlgoliaPlaces/AlgoliaPlaces.php [hunk 2]
```diff
@@ -72,20 +75,33 @@ public function getName(): string
...
             return new AddressCollection([]);
         }
-
-        return $this->buildResult($jsonResponse);
+        if ($jsonResponse['degradedQuery']) {
+            return new AddressCollection([]);
+        }
+        if ($jsonResponse['nbHits'] == 0) {
+            return new AddressCollection([]);
+        } else {
```
Comment (Fabex) [inline src/Provider/AlgoliaPlaces/AlgoliaPlaces.php]: and no need else

File: src/Provider/AlgoliaPlaces/AlgoliaPlaces.php [hunk 3]
```diff
@@ -110,7 +126,35 @@ protected function getRequest(string $url): RequestInterface
...
+    /**
+     * Send request and return contents. If content is empty, an exception will be thrown.
+     *
+     * @param RequestInterface $request
+     *
+     * @return string
+     *
+     * @throws InvalidServerResponse
+     */
+    protected function getParsedResponse(RequestInterface $request): string
```
Comment (Fabex) [inline src/Provider/AlgoliaPlaces/AlgoliaPlaces.php]: still weird for me to have to rewrite this method

**Review by Fabex — CHANGES_REQUESTED** [review]
Just some little change.
but big thx for this job

File: src/Provider/AlgoliaPlaces/AlgoliaPlaces.php (hunk 3 above)
Comment (sheub) [reply src/Provider/AlgoliaPlaces/AlgoliaPlaces.php]: I agree,
Feel free to try it yourself: 
the tests at home passed alright (php7.2 and php7.0) without the getParsedResponse function, but Travis throw me an error if it is not there.

**Comment by sheub** [comment]
maybe you can check the travis build here:
https://travis-ci.org/sheub/Geocoder/builds/404961176

**Comment by sheub** [comment]
I don't know if you have access to my travis builds?
The last commit failed the Tests because the query is not set correctly

**Comment by sheub** [comment]
What do you think?

**Comment by sheub** [comment]
Here: I call explicitly AbstractHttpProvider::getParsedResponse() from AlgoliaPlaces.php, and Travis throw me:
`Error: Call to undefined method Geocoder\Http\Provider\AbstractHttpProvider::getParsedResponse()`

**Comment by sheub** [comment]
ping @Fabex 
Do you prefer a version with the `getParsedResponse` function which pass the build or a version without the function?
Either way, will have to clear this up with Nyholm.

**Comment by sheub** [comment]
I'm just closing because of inactivity

**Comment by Fabex** [comment]
yes sorry I was on vacation and I start a new job so I was not a lot of time for OS.
I will take a look soon

**Comment by sheub** [comment]
Hi @Fabex ,
I have reopened this PR. I would be good if you could merge it as soon as you can.
If you really are to busy and don't have enough time for it, just let me know.

Please don't just leave it like this.

Thanks

**Review by Fabex — APPROVED** [review]

**Comment by Fabex** [comment]
hi @sheub 
I merged your PR.
But I have some failed tests. I think I don't really understand how integrations tests works.
I saw that there is several cached responses, but the first time I ran the integrations test, a new cached response was created and each other time I ran integrations tests it use the same new cached response and not yours. 
Can you help me to fix that, please ?

**Comment by sheub** [comment]
Hi @Fabex 
Thx for mergin. What I can see from the travis build is 
`Error: Call to undefined method Geocoder\Http\Provider\AbstractHttpProvider::getParsedResponse()`

I don't know how to resolve this error without adding a copy of the `getParsedResponse()` function directly to the `AlgoliaPlaces` class.

**Comment by sheub** [comment]
see https://github.com/Fabex/Geocoder/pull/1#issue-202000796

**Comment by Fabex** [comment]
@sheub In fact in local, I don't have issue about ```getParsedResponse()``` but about cachedResponse,
below my phpunit stacktrace

```
fabex: ~/dev/Geocoder/src/Provider/AlgoliaPlaces (feature/algolia-places-provider⚡)$./vendor/bin/phpunit 
PHPUnit 6.3.1 by Sebastian Bergmann and contributors.

......FFSSSS.....                                                 17 / 17 (100%)

Time: 85 ms, Memory: 6.00MB

There were 2 failures:

1) Geocoder\Provider\AlgoliaPlaces\Tests\IntegrationTest::testGeocodeQuery
Latitude should be in London
Failed asserting that 48.8679 matches expected 51.5033.

/home/fabex/dev/Geocoder/src/Provider/AlgoliaPlaces/Tests/IntegrationTest.php:108

2) Geocoder\Provider\AlgoliaPlaces\Tests\IntegrationTest::testGeocodeQueryWithNoResults
Failed asserting that 1 matches expected 0.

/home/fabex/dev/Geocoder/src/Provider/AlgoliaPlaces/Tests/IntegrationTest.php:130

FAILURES!
Tests: 17, Assertions: 49, Failures: 2, Skipped: 4.
```

**Comment by sheub** [comment]
Hi @Fabex ,
The tests will run on the cached_responses if they are there (the caching have been added to save geocoding requests to the provider).
I can't tell for now what's wrong. 
You can try to delete the cached_responses and rerun the tests twice.
-> do you have both run failing?

**Comment by Fabex** [comment]
Hi @sheub,
I think I have found a solution but I should modified ```geocoder-php/provider-integration-tests/src/CachedResponseClient.php```.
Because, the cache was generated like this
```
$file = sprintf('%s/%s_%s', $this->cacheDir, $host, sha1($url));
```
and as you can see the file name is only based on the url of the query
but AlgoliaPlace send query in the body of the post request. So if for each query the url are the same and generate all time the same cached file no matter the query data.
so for the moment I try this
```
$file = sprintf('%s/%s_%s_%s', $this->cacheDir, $host, sha1($url), sha1($body));
```
and now the cached file name is different for each different query.

and all my tests are in success.

**Comment by Fabex** [comment]
@sheub ok forget my last message my vendors was not up to date and I was not the last version of ```geocoder-php/provider-integration-tests``` so no there is no problem.

**Pull request merged**